testing/trino-tests/src/test/java/io/trino/connector/system/metadata/TestSystemMetadataConnector.java

```diff
@@ -0,0 +1,385 @@
+                        .build());
+
+        // Non-existent catalog
+        assertMetadataCalls(
+                "SELECT comment FROM system.metadata.materialized_views WHERE catalog_name = 'wrong'",
+                "SELECT '' WHERE false",
+                ImmutableMultiset.of());
+
+        // Empty catalog name
+        assertMetadataCalls(
+                "SELECT comment FROM system.metadata.materialized_views WHERE catalog_name = ''",
+                "SELECT '' WHERE false",
+                ImmutableMultiset.of());
+
+        // Empty table schema and table name TODO should not fail
+        assertQueryFails(
+                "SELECT comment FROM system.metadata.materialized_views WHERE schema_name = '' AND name = ''",
+                "Error listing materialized views for catalog \\w+: schemaName is empty");
+
+        // Empty table schema TODO should not fail
+        assertThatThrownBy(() -> computeActual("SELECT count(comment) FROM system.metadata.materialized_views WHERE schema_name = ''"))
+                .isInstanceOf(QueryFailedException.class)
+                .hasMessageMatching("Error listing materialized views for catalog \\w+: schemaName is empty");
+
+        // Empty table name
+        assertQueryFails(
+                "SELECT count(comment) FROM system.metadata.materialized_views WHERE name = ''",
+                "Error listing materialized views for catalog broken_catalog: Catalog is broken");
+    }
+
+    @Test
+    public void testMetadataListingExceptionHandling()
+    {
+        // TODO this should probably gracefully continue when some catalog is "broken" (does not currently work, e.g. is offline)
+        assertQueryFails(
+                "SELECT * FROM system.metadata.table_comments",
+                "Catalog is broken");
+
+        // TODO this should probably gracefully continue when some catalog is "broken" (does not currently work, e.g. is offline)
+        assertQueryFails(
+                "SELECT * FROM system.metadata.materialized_views",
+                "Error listing materialized views for catalog broken_catalog: Catalog is broken");
+    }
+
+    private void assertMetadataCalls(@Language("SQL") String actualSql, @Language("SQL") String expectedSql, Multiset<String> expectedMetadataCallsCount)
+    {
+        Multiset<String> actualMetadataCallsCount = countingMockConnector.runTracing(() -> {
+            // expectedSql is run on H2, so does not affect counts.
+            assertQuery(actualSql, expectedSql);
+        });
+
+        actualMetadataCallsCount = actualMetadataCallsCount.stream()
+                // Every query involves beginQuery and cleanupQuery, so ignore them.
+                .filter(method -> !"ConnectorMetadata.beginQuery".equals(method) && !"ConnectorMetadata.cleanupQuery".equals(method))
+                .collect(toImmutableMultiset());
+
+        assertMultisetsEqual(actualMetadataCallsCount, expectedMetadataCallsCount);
+    }
+}
```

Comment:
Add tests where you want to get objects from `broken_catalog` explicitely

Reply:
let's follow-up on this!